{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# pip install pygenomeviz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pygenomeviz import GenomeViz\n",
    "import random\n",
    "import sys\n",
    "\n",
    "# random.seed(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# process minimap alignment\n",
    "home=\"/Users/luorunpeng/Downloads/all-e/Research/project-haplotype-phasing/ctgContact-dup\"\n",
    "chrom_file=home+\"/chrom_id_pst104e.lst\"\n",
    "\n",
    "ref_file=\"/Users/luorunpeng/Downloads/all-e/Research/project-haplotype-phasing/run_pst104e/reference/chrAll.fna\"\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# parse chrom file\n",
    "chr_id2name = {}\n",
    "name2chr_id = dict()\n",
    "lefts = []\n",
    "rights = []\n",
    "with open(chrom_file, \"r\") as chrom_fd:\n",
    "    for line in chrom_fd:\n",
    "        chrom_id, left, right = line.strip().split(\"\\t\")\n",
    "        chr_id2name[chrom_id] = (left, right)\n",
    "        name2chr_id[left] = chrom_fd\n",
    "        name2chr_id[right] = chrom_fd\n",
    "        lefts.append(left)\n",
    "        rights.append(right)\n",
    "    chrom_fd.close()\n",
    "\n",
    "def get_strand(strand: str):\n",
    "    if strand == \"+\":\n",
    "        return 1\n",
    "    else:\n",
    "        return -1\n",
    "\n",
    "def parse_aln_file(aln_file: str):\n",
    "    chrom_dict = {}\n",
    "    for name in name2chr_id.keys():\n",
    "        chrom_dict[name] = []\n",
    "    with open(aln_file, \"r\") as paf_fd:\n",
    "        for line in paf_fd:\n",
    "            # perfect alignment\n",
    "            splited = line.strip().split(\"\\t\")\n",
    "            seg_no = str(splited[0])\n",
    "            strand = get_strand(str(splited[4]))\n",
    "            r_name = str(splited[5])\n",
    "            r_len = int(splited[6])\n",
    "            r_start, r_end = int(splited[7]), int(splited[8])\n",
    "            mapq = int(splited[11])\n",
    "            chrom_dict[r_name].append((seg_no, strand, mapq, r_start, r_end))\n",
    "        paf_fd.close()\n",
    "    return chrom_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_rlens(ref_file: str):\n",
    "    res = {}\n",
    "    with open(ref_file, \"r\") as fa_fd:\n",
    "        sid = \"\"\n",
    "        seq = \"\"\n",
    "        for line in fa_fd:\n",
    "            if line.startswith(\">\"):\n",
    "                # process previous entry\n",
    "                if sid != \"\":\n",
    "                    res[sid] = len(seq)\n",
    "                sid = line.strip()[1:].split()[0]\n",
    "                seq = \"\"\n",
    "            else:\n",
    "                seq += line.strip()\n",
    "        fa_fd.close()\n",
    "        if sid != \"\":\n",
    "            res[sid] = len(seq)\n",
    "    return res\n",
    "\n",
    "# parse the Quast csv file\n",
    "def parse_tsv_file(tsv_file: str):\n",
    "    chrom_dict = {}\n",
    "    for name in name2chr_id.keys():\n",
    "        chrom_dict[name] = []\n",
    "    with open(tsv_file, \"r\") as tsv_fd:\n",
    "        tsv_fd.readline()\n",
    "        i = 0\n",
    "        for line in tsv_fd.readlines():\n",
    "            if i == 0:\n",
    "                # line 0\n",
    "                if \"unaligned\" in line:\n",
    "                    continue\n",
    "                splited = line.strip().split()\n",
    "                s1, e1 = int(splited[0]), int(splited[1])\n",
    "                rname = str(splited[4])\n",
    "                cname = str(splited[5])\n",
    "                mrate = float(splited[6])\n",
    "                chrom_dict[rname].append((cname, -2, mrate, s1, e1))\n",
    "            else:\n",
    "                # line 1\n",
    "                pass # contig alignment type\n",
    "            \n",
    "            i = (i+1) % 2\n",
    "        tsv_fd.close()\n",
    "    return chrom_dict\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "def draw_track(pts: list, track, strand, fg_color, add_label):\n",
    "    for cds in pts:\n",
    "        (seg_no, _, mapq, r_start, r_end) = cds\n",
    "        # strand = 1 # ignore the strand when plot\n",
    "        plotstyle = \"box\"\n",
    "        # random.choice(plotstyles)\n",
    "        color = fg_color\n",
    "\n",
    "        # label=seg_no, labelsize=10, \n",
    "        if add_label:\n",
    "            track.add_feature(r_start, r_end, strand, facecolor=color, plotstyle=plotstyle, label=seg_no, labelsize=10, labelrotation=90, linewidth=0.5)\n",
    "        else:\n",
    "            track.add_feature(r_start, r_end, strand, facecolor=color, plotstyle=plotstyle, linewidth=0.5)\n",
    "        # track.add_feature(start, end, strand, label=plotstyle, labelsize=20, facecolor=color, plotstyle=plotstyle, labelvpos=\"top\", labelhpos=\"center\", labelrotation=60, linewidth=1)\n",
    "    return\n",
    "\n",
    "def draw_phasing_plot(tsv_file1: str, tsv_file2: str, ref_file: str, output):\n",
    "    gv = GenomeViz()\n",
    "\n",
    "    chrom_dict1 = parse_tsv_file(tsv_file1)\n",
    "    chrom_dict2 = parse_tsv_file(tsv_file2)\n",
    "    chrom_len = get_rlens(ref_file)\n",
    "\n",
    "    plotstyles = (\"bigarrow\", \"arrow\", \"bigbox\", \"box\", \"bigrbox\", \"rbox\")\n",
    "    colors = (\"orange\", \"blue\", \"lime\", \"red\")\n",
    "\n",
    "    color1=\"blue\"\n",
    "    color2=\"red\"\n",
    "\n",
    "    # gv = GenomeViz()\n",
    "    for chr_id, (left, right) in chr_id2name.items():\n",
    "\n",
    "        # if chr_id != '3':\n",
    "        #     continue\n",
    "\n",
    "        left_size, right_size = chrom_len[left], chrom_len[right]\n",
    "        max_size = max(left_size, right_size)\n",
    "\n",
    "        # l_track = gv.add_feature_track(name, max_size)\n",
    "        l_track = gv.add_feature_track(f\"{chr_id}A\\n{chr_id}B\", max_size)\n",
    "        l_track.add_feature(0, max_size, 1, facecolor=\"grey\", plotstyle=\"box\", linewidth=1)\n",
    "        \n",
    "        draw_track(chrom_dict1[left], l_track, 1, color1, False)\n",
    "        draw_track(chrom_dict1[right], l_track, 1, color2, False)\n",
    "\n",
    "        l_track.add_feature(0, max_size, -1, facecolor=\"grey\", plotstyle=\"box\", linewidth=1)\n",
    "\n",
    "        draw_track(chrom_dict2[right], l_track, -1, color2, False)\n",
    "        draw_track(chrom_dict2[left], l_track, -1, color1, False)\n",
    "\n",
    "\n",
    "    # fig = gv.plotfig(dpi=300)\n",
    "    gv.savefig(output, dpi=300)\n",
    "    return\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "quast_base=\"/Users/luorunpeng/Downloads/all-e/Research/project-haplotype-phasing/quast_analysis\"\n",
    "quast_dir_all=quast_base+\"/quast_pst104e_all/contigs_reports\"\n",
    "quast_dir_hapA=quast_base+\"/quast_pst104e_hapA/contigs_reports\"\n",
    "quast_dir_hapB=quast_base+\"/quast_pst104e_hapB/contigs_reports\"\n",
    "\n",
    "# hifiasm (hic)\n",
    "h1_hifiasm_hic_tsv=quast_dir_hapA+\"/all_alignments_pst104e_ul_hic-asm-hic-hap2-p_ctg.tsv\"\n",
    "h2_hifiasm_hic_tsv=quast_dir_hapB+\"/all_alignments_pst104e_ul_hic-asm-hic-hap1-p_ctg.tsv\"\n",
    "\n",
    "draw_phasing_plot(h1_hifiasm_hic_tsv, h2_hifiasm_hic_tsv, ref_file, home+\"/imgs/pst104e_hifiasm_hic_phasing_plot.png\")\n",
    "\n",
    "# hifiasm + greenhill\n",
    "h1_hifiasm_greenhill_tsv=quast_dir_hapA+\"/all_alignments_greenhill_hifiasm_H2.tsv\"\n",
    "h2_hifiasm_greenhill_tsv=quast_dir_hapB+\"/all_alignments_greenhill_hifiasm_H1.tsv\"\n",
    "\n",
    "draw_phasing_plot(h1_hifiasm_greenhill_tsv, h2_hifiasm_greenhill_tsv, ref_file, home+\"/imgs/pst104e_hifiasm_greenhill_phasing_plot.png\")\n",
    "\n",
    "# hifiasm + algorithm\n",
    "h1_hifiasm_tree_tsv=quast_dir_hapA+\"/all_alignments_dir_phasing_hifiasm_pst104e_tree_bin1.tsv\"\n",
    "h2_hifiasm_tree_tsv=quast_dir_hapB+\"/all_alignments_dir_phasing_hifiasm_pst104e_tree_bin0.tsv\"\n",
    "\n",
    "draw_phasing_plot(h1_hifiasm_tree_tsv, h2_hifiasm_tree_tsv, ref_file, home+\"/imgs/pst104e_tree_hifiasm_phasing_plot.png\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "quast_base=\"/Users/luorunpeng/Downloads/all-e/Research/project-haplotype-phasing/quast_analysis\"\n",
    "quast_dir_all=quast_base+\"/quast_pst104e_all/contigs_reports\"\n",
    "quast_dir_hapA=quast_base+\"/quast_pst104e_hapA/contigs_reports\"\n",
    "quast_dir_hapB=quast_base+\"/quast_pst104e_hapB/contigs_reports\"\n",
    "\n",
    "# verkko (hic)\n",
    "h1_verkko_hic_tsv=quast_dir_hapA+\"/all_alignments_assembly-haplotype1.tsv\"\n",
    "h2_verkko_hic_tsv=quast_dir_hapB+\"/all_alignments_assembly-haplotype2.tsv\"\n",
    "\n",
    "draw_phasing_plot(h1_verkko_hic_tsv, h2_verkko_hic_tsv, ref_file, home+\"/imgs/pst104e_verkko_hic_phasing_plot.png\")\n",
    "\n",
    "\n",
    "\n",
    "# verkko + greenhill\n",
    "h1_verkko_greenhill_tsv=quast_dir_hapA+\"/all_alignments_greenhill_verkko_H1.tsv\"\n",
    "h2_verkko_greenhill_tsv=quast_dir_hapB+\"/all_alignments_greenhill_verkko_H2.tsv\"\n",
    "\n",
    "draw_phasing_plot(h1_verkko_greenhill_tsv, h2_verkko_greenhill_tsv, ref_file, home+\"/imgs/pst104e_verkko_greenhill_phasing_plot.png\")\n",
    "\n",
    "\n",
    "\n",
    "# verkko + algorithm\n",
    "h1_verkko_tree_tsv=quast_dir_hapA+\"/all_alignments_dir_phasing_verkko_pst104e_tree_bin1.tsv\"\n",
    "h2_verkko_tree_tsv=quast_dir_hapB+\"/all_alignments_dir_phasing_verkko_pst104e_tree_bin0.tsv\"\n",
    "\n",
    "draw_phasing_plot(h1_verkko_tree_tsv, h2_verkko_tree_tsv, ref_file, home+\"/imgs/pst104e_tree_verkko_phasing_plot.png\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "draw_phasing_plot(h1_hifiasm_hic_tsv, h2_hifiasm_hic_tsv, ref_file, home+\"/pt76_hifiasm_hic_phasing_plot.png\", \"Hifiasm (HiC)\")\n",
    "draw_phasing_plot(h1_hifiasm_greenhill_tsv, h2_hifiasm_greenhill_tsv, ref_file, home+\"/pt76_hifiasm_greenhill_phasing_plot.png\", \"Hifiasm + GreenHill\")\n",
    "draw_phasing_plot(h1_tree_hifiasm_hic_tsv, h2_tree_hifiasm_hic_tsv, ref_file, home+\"/pt76_tree_hifiasm_hic_phasing_plot.png\", \"Hifiasm+Algorithm\")\n",
    "draw_phasing_plot(h1_verkko_hic_tsv, h2_verkko_hic_tsv, ref_file, home+\"/pt76_verkko_phasing_plot.png\", \"Verkko (HiC)\")\n",
    "draw_phasing_plot(h1_verkko_greenhill_tsv, h2_verkko_greenhill_tsv, ref_file, home+\"/pt76_verkko_greenhill_phasing_plot.png\", \"Verkko + GreenHill\")\n",
    "draw_phasing_plot(h1_tree_verkko_tsv, h2_tree_verkko_tsv, ref_file, home+\"/pt76_tree_verkko_phasing_plot.png\", \"Verkko + Algorithm\")\n",
    "\n",
    "gv.savefig(\"pt76_chr3_phasing_plot.png\", dpi=300)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "bio-env",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.17"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
